src/storage/sql/migrations/versions/1bf4e4e7b24f_initial_schema.py

```diff
@@ -0,0 +1,103 @@
+
+        # slugs must be unique (this will also automatically create a unique index)
+        sa.UniqueConstraint('slug', name='uq_slug'),
+
+        # will be used for pagination; slug guarantees uniqueness of the sorting key
+        sa.Index('snippets_created_at_slug', 'created_at', 'slug'),
+        sa.Index('snippets_updated_at_slug', 'updated_at', 'slug'),
+    )
+
+    op.create_table(
+        'changesets',
+
+        sa.Column('id', sa.Integer, primary_key=True),
+        sa.Column('snippet_id', sa.Integer, nullable=False),
+
+        # numeric index used to determine the ordering of changesets for a given snippet
+        sa.Column('version', sa.Integer,
+                  server_default=sa.text('0'), nullable=False),
+        sa.Column('content', sa.Text, nullable=False),
+        sa.Column('created_at', sa.DateTime(timezone=True),
+                  server_default=sa.func.now(), nullable=False),
+        sa.Column('updated_at', sa.DateTime(timezone=True),
+                  server_default=sa.func.now(), nullable=False),
+
+        # each changeset is supposed to have a unique version number
+        sa.UniqueConstraint('snippet_id', 'version',
+                            name='uq_version'),
+        # sanity check: do not allow empty changesets
+        sa.CheckConstraint('LENGTH(content) > 0',
+                           name='check_not_empty'),
+        # sanity check: version numbers are non-negative integers
+        sa.CheckConstraint('version >= 0',
+                           name='check_non_negative_version')
+    )
+    with op.batch_alter_table('changesets') as batch_op:
+        # there can be multiple changesets per snippet; changesets should be
+        # deleted when the parent snippet is deleted
+        batch_op.create_foreign_key('fk_snippet', 'snippets',
+                                    ['snippet_id'], ['id'],
+                                    ondelete='CASCADE')
+
+    # tags could have been associated with snippets as M:M via an auxiliary table,
+    # but Diesel only supports child-parent associations, so let's do that instead
+    op.create_table(
+        'tags',
+
+        sa.Column('id', sa.Integer, primary_key=True),
+        sa.Column('snippet_id', sa.Integer, nullable=False),
+
+        sa.Column('value', sa.Text, nullable=False),
+
+        # do not allow to abuse the tags for storing too much data
+        sa.CheckConstraint('LENGTH(value) < 128', name='check_length'),
+        # do not allow repeated tags per snippet
+        sa.UniqueConstraint('snippet_id', 'value', name='uq_snippet_tag'),
+    )
+    with op.batch_alter_table('tags') as batch_op:
+        # there can be multiple tags per snippet; tags should be deleted when
+        # the parent snippet is deleted
+        batch_op.create_foreign_key('fk_snippet', 'snippets',
+                                    ['snippet_id'], ['id'],
+                                    ondelete='CASCADE')
+
+
+def downgrade():
+    op.drop_table('tags')
```

Comment:
Just to double confirm, you decided that we don't want to go with intermediate `snippets_tags` table that glues together `tags` and  `snippets`?

Reply:
Correct. In order to keep things simpler on the querying side, as apparently Diesel DSL is a bit picky about multiple JOINs in the same query.